File: src/Models/Buildable/Installable/DoubleStud.java
package Models.Buildable.Installable;

import Graphics.GraphicsList;
import Graphics.RectangleInstructions;
import Models.Buildable.Material.Lumber;
import Models.Measurement;

/**
 * Frequently, studs are place right next to each other and that requires them to be nailed together. This class ensures
 * the material list and drawing instructions are updated accordingly.
 */
public class DoubleStud extends Stud {
    /**
     *
     * @param length - Length of Double stud to create
     * @param dimension - Dimension of one of the studs to create the doubleStud
     */
    public DoubleStud(Measurement length, Lumber.Dimension dimension) {
        super(length, dimension);

        // The second stud is attached like a plate, so use the static method here.
        this.material.addMaterial(Stud.nailType, Plate.numberOfNails(length));
        this.material.addMaterial(new Lumber(length, dimension), 1);
    }

    /**
     * A Convenience constructor for making double studs from a stud
     * @param stud - Stud to create a double stud from
     */
    public DoubleStud(Stud stud) {
        this(stud.installedHeight, stud.dimension);
    }


    /**
     *
     * @return - A copy of the {@link Measurement} of the width of this DoubleStud
     */
    @Override
    public Measurement totalWidth() {
        return this.dimension.width.clone().multiply(2);
    }

    /**
     *
     * @return - A {@link GraphicsList} of {@link Graphics.GraphicsInstructions} used to draw this DoubleStud
     */
    @Override
    public GraphicsList graphicsList() {
        GraphicsList result = super.graphicsList();
        Measurement zero = new Measurement(0);
        return result.addGraphic(new RectangleInstructions(this.dimension.width, zero, this.dimension.width, this.installedHeight));
    }
}

File: src/main/kotlin/Models/buildable/installable/DoubleStud.kt
package models.buildable.installable

import graphics.GraphicsList
import graphics.RectangleInstructions
import models.buildable.material.Lumber
import models.Measurement

/**
 * Frequently, studs are place right next to each other and that requires them to be nailed together. This class ensures
 * the material list and drawing instructions are updated accordingly.
 * @param length - Length of Double stud to create
 * @param dimension - Dimension of one of the studs to create the doubleStud
 */
open class DoubleStud(length: Measurement, dimension: Lumber.Dimension) : Stud(length, dimension) {
    init {
        // The second stud is attached like a plate, so use the Plate static method here.
        material.addMaterial(nailType, Plate.numberOfNails(length))
        material.addMaterial(Lumber(length, dimension), 1)
    }

    /**
     * A Convenience constructor for making double studs from a stud
     * @param stud - Stud to create a double stud from
     */
    constructor(stud: Stud) : this(stud.totalHeight(), stud.dimension)

    /**
     *
     * @return - A copy of the [Measurement] of the width of this DoubleStud
     */
    override fun totalWidth(): Measurement {
        return super.totalWidth().multiply(2)
    }

    /**
     *
     * @return - A [GraphicsList] of [graphics.GraphicsInstructions] used to draw this DoubleStud
     */
    override fun graphicsList(): GraphicsList {
        val result = super.graphicsList()
        val zero = Measurement(0)
        return result.addGraphic(RectangleInstructions(super.totalWidth(), zero, super.totalWidth(), totalHeight()))
    }
}